File: part-02/src/test/java/eu/luminis/workshop/smallsteps/logic/domainservice/LegoStockReturningInCompleteTest.java
package eu.luminis.workshop.smallsteps.logic.domainservice;

import eu.luminis.workshop.smallsteps.logic.domainmodel.valueobjects.LegoSetNumber;
import eu.luminis.workshop.smallsteps.logic.domainservice.helper.SetupLegoTestApp;
import eu.luminis.workshop.smallsteps.logic.domainservice.state.IncompleteReturn;
import eu.luminis.workshop.smallsteps.logic.domainservice.state.LegoBox;
import eu.luminis.workshop.smallsteps.logic.domainservice.state.LegoSetForBuilder;
import eu.luminis.workshop.smallsteps.logic.domainservice.state.StockState;
import eu.luminis.workshop.smallsteps.persistence.InMemoryLegoPartCatalog;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import java.util.Collections;
import java.util.List;
import java.util.Map;

import static org.assertj.core.api.Assertions.assertThatThrownBy;
import static org.assertj.core.api.AssertionsForInterfaceTypes.assertThat;

public class LegoStockReturningInCompleteTest {
    SetupLegoTestApp app;
    Map<LegoSetNumber, Map<String, Integer>> partsForLegoSets;
    Map<String, Integer> missingPartsForMilleniumFalcon;
    Map<String, Integer> nonMilleniumFalconParts;


    @BeforeEach
    void setUp() {
        app = new SetupLegoTestApp();
        partsForLegoSets = Map.of(
                app.millenniumFalcon, Map.of(
                        "3022", 8,
                        "2420", 2,
                        "60581", 2,
                        "4865b", 2,
                        "20105", 1
                )
        );
        missingPartsForMilleniumFalcon = Map.of("20105", 1, "3022", 3);
        nonMilleniumFalconParts = Map.of("00000", 1,"99999c",2);
    }

    @Test
    public void should_register_returning_incomplete_lego_set() {
        StockState initialState = new StockState(
                null,
                null,
                List.of(new LegoSetForBuilder(app.millenniumFalcon, app.harry)),
                null,
                null
        );
        LegoStock handler = new LegoStock(app.harryAuth, new InMemoryLegoPartCatalog(partsForLegoSets), initialState);

        StockState newState = handler.returnInComplete(
                new LegoBox(app.millenniumFalcon, 42, missingPartsForMilleniumFalcon));

        assertThat(List.of(new LegoBox(app.millenniumFalcon, 42, missingPartsForMilleniumFalcon)))
                .isEqualTo(newState.getIncompleteStock());
        assertThat(List.of(new IncompleteReturn(app.millenniumFalcon, missingPartsForMilleniumFalcon)))
                .isEqualTo(newState.getIncompleteReturnHistory());

        assertThat(newState.getAtBuilder()).isEqualTo(initialState.getAtBuilder());
        assertThat(newState.getAvailableStock()).isEqualTo(initialState.getAvailableStock());
        assertThat(newState.getReserved()).isEqualTo(initialState.getReserved());
    }

    @Test
    public void should_fail_to_register_incomplete_lego_set_when_no_missing_parts_specified() {
        StockState initialState = new StockState(
                null,
                null,
                List.of(new LegoSetForBuilder(app.millenniumFalcon, app.harry)),
                null,
                null
        );
        LegoStock handler = new LegoStock(app.harryAuth, new InMemoryLegoPartCatalog(partsForLegoSets), initialState);

        assertThatThrownBy(() -> {
            handler.returnInComplete(new LegoBox(app.millenniumFalcon, 42, Collections.emptyMap()));
        }).isInstanceOf(IllegalArgumentException.class);
    }

    @Test
    public void should_fail_to_register_incomplete_lego_set_when_missing_parts_specified_do_not_belong_to_legoset() {
        StockState initialState = new StockState(
                null,
                null,
                List.of(new LegoSetForBuilder(app.millenniumFalcon, app.harry)),
                null,
                null
        );
        LegoStock handler = new LegoStock(app.harryAuth, new InMemoryLegoPartCatalog(partsForLegoSets), initialState);

        assertThatThrownBy(() -> {
            handler.returnInComplete(new LegoBox(app.millenniumFalcon, 42, nonMilleniumFalconParts));
        }).isInstanceOf(IllegalArgumentException.class);

    }

    @Test
    public void should_fail_to_register_incomplete_lego_set_when_missing_parts_specified_more_than_in_set() {
        StockState initialState = new StockState(
                null,
                null,
                List.of(new LegoSetForBuilder(app.millenniumFalcon, app.harry)),
                null,
                null
        );
        LegoStock handler = new LegoStock(app.harryAuth, new InMemoryLegoPartCatalog(partsForLegoSets), initialState);

        assertThatThrownBy(() -> {
            handler.returnInComplete(new LegoBox(app.millenniumFalcon, 42, Map.of("20105",5)));
        }).isInstanceOf(IllegalArgumentException.class);

        assertThatThrownBy(() -> {
            handler.returnInComplete(new LegoBox(app.millenniumFalcon, 42, Map.of("20105",0)));
        }).isInstanceOf(IllegalArgumentException.class);

    }

}

File: part-02/src/test/kotlin/eu/luminis/workshop/smallsteps/logic/domainService/LegoStock - ReturningInCompleteTest.kt
package eu.luminis.workshop.smallsteps.logic.domainService

import eu.luminis.workshop.smallsteps.logic.domainService.helper.LegoStockHelper
import eu.luminis.workshop.smallsteps.logic.domainService.helper.LegoStockHelper.millenniumFalcon
import eu.luminis.workshop.smallsteps.logic.domainService.helper.TestUsers
import eu.luminis.workshop.smallsteps.logic.domainService.state.IncompleteReturn
import eu.luminis.workshop.smallsteps.logic.domainService.state.LegoBox
import eu.luminis.workshop.smallsteps.logic.domainService.state.StockState
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFailsWith

@Suppress("ClassName")
internal class `LegoStock - ReturningInCompleteTest` {
    private val partCatalog = mapOf(
        // https://www.bricklink.com/catalogItemInv.asp?S=75192-1
        millenniumFalcon to mapOf(
            "3022" to 8, // Black Plate 2 x 2
            "2420" to 2, // Black Plate 2 x 2 Corner
            "60581" to 2, // Black Panel 1 x 4 x 3 with Side Supports - Hollow Studs
            "4865b" to 2, // Black Panel 1 x 2 x 1 with Rounded Corners
            "20105" to 1 // Black Minifigure, Weapon Crossbow with Mini Blaster / Shooter
        )
    )

    private val missingPartsForMilleniumFalcon = mapOf(
        "20105" to 1,
        "3022" to 3,
    )

    private val nonMilleniumFalconParts = mapOf(
        "00000" to 1,
        "99999c" to 2,
    )

    @Test
    fun `should register returning incomplete lego set`() {
        // given
        val initialState = StockState(
            atBuilder = listOf(
                millenniumFalcon to TestUsers.harry
            ),
        )
        val handler = LegoStockHelper.build(
            initialState,
            partCatalog = partCatalog,
        )

        // when
        val newState = handler.returningInComplete(LegoBox(millenniumFalcon, 42, missingPartsForMilleniumFalcon))

        // then
        assertEquals(listOf(LegoBox(millenniumFalcon, 42, missingPartsForMilleniumFalcon)), newState.incompleteStock)
        assertEquals(
            listOf(IncompleteReturn(millenniumFalcon, missingPartsForMilleniumFalcon)),
            newState.incompleteReturnHistory
        )
        assertEquals(initialState.atBuilder, newState.atBuilder)
        assertEquals(initialState.availableStock, newState.availableStock)
        assertEquals(initialState.reserved, newState.reserved)
    }

    @Test
    fun `should fail to register returning incomplete lego set when no missing parts specified`() {
        // given
        val initialState = StockState(
            atBuilder = listOf(
                millenniumFalcon to TestUsers.harry
            ),
        )
        val handler = LegoStockHelper.build(
            initialState,
            partCatalog = partCatalog,
        )

        // then
        assertFailsWith<IllegalArgumentException> {
            // when
            handler.returningInComplete(LegoBox(millenniumFalcon, 42, emptyMap()))
        }
    }

    @Test
    fun `should fail to register returning incomplete lego set when missing parts specified that don't belong to the lego set`() {
        // given
        val initialState = StockState(
            atBuilder = listOf(
                millenniumFalcon to TestUsers.harry
            ),
        )
        val handler = LegoStockHelper.build(
            initialState,
            partCatalog = partCatalog,
        )

        // then
        assertFailsWith<IllegalArgumentException> {
            // when
            handler.returningInComplete(LegoBox(millenniumFalcon, 42, nonMilleniumFalconParts))
        }

        // and
        assertFailsWith<IllegalArgumentException> {
            // when
            handler.returningInComplete(
                LegoBox(
                    millenniumFalcon,
                    42,
                    nonMilleniumFalconParts + nonMilleniumFalconParts
                )
            )
        }
    }

    @Test
    fun `should fail to register returning incomplete lego set when more missing parts are reported than what was in the original set`() {
        // given
        val initialState = StockState(
            atBuilder = listOf(
                millenniumFalcon to TestUsers.harry
            ),
        )
        val handler = LegoStockHelper.build(
            initialState,
            partCatalog = partCatalog,
        )

        // then
        assertFailsWith<IllegalArgumentException> {
            // when
            handler.returningInComplete(LegoBox(millenniumFalcon, 42, mapOf("20105" to 5))) // only 1 part of set
        }

        // and
        assertFailsWith<IllegalArgumentException> {
            // when
            handler.returningInComplete(LegoBox(millenniumFalcon, 42, mapOf("20105" to 0)))
        }
    }

    // count total missing parts

}